client-common/src/tendermint/rpc_client/async_rpc_client.rs

```diff
@@ -0,0 +1,262 @@
+    ///
+    /// # Note
+    ///
+    /// This does not use batch JSON-RPC requests but makes multiple single JSON-RPC requests in parallel.
+    ///
+    /// TODO: Usage of `Vec` can be removed once we execute it in a purely async context
+    pub async fn request_batch(&self, batch_params: &[(&str, Vec<Value>)]) -> Result<Vec<Value>> {
+        if batch_params.is_empty() {
+            // Do not send empty batch requests
+            return Ok(Default::default());
+        }
+
+        let mut receivers = Vec::with_capacity(batch_params.len());
+
+        for (ref method, ref params) in batch_params.iter() {
+            let (id, channel_receiver) = self.send_request(method, params).await?;
+            receivers.push((id, channel_receiver));
+        }
+
+        let mut responses = Vec::with_capacity(batch_params.len());
+
+        for (i, (id, receiver)) in receivers.into_iter().enumerate() {
+            let method = batch_params[i].0;
+            let params = &batch_params[i].1;
+
+            let response = self.receive_response(method, params, &id, receiver).await?;
+            responses.push(response);
+        }
+
+        Ok(responses)
+    }
+
+    /// Makes an RPC call and deserializes the response
+    pub async fn call<T>(&self, method: &str, params: &[Value]) -> Result<T>
+    where
+        for<'de> T: Deserialize<'de>,
+    {
+        let response_value = self.request(method, params).await?;
+        serde_json::from_value(response_value).with_context(|| {
+            format!(
+                "Unable to deserialize `{}` from JSON-RPC response for params: {:?}",
+                method, params
+            )
+        })
+    }
+
+    /// Makes RPC call in batch and deserializes responses
+    ///
+    /// TODO: Usage of `Vec` can be removed once we execute it in a purely async context
+    pub async fn call_batch<T>(&self, batch_params: &[(&str, Vec<Value>)]) -> Result<Vec<T>>
+    where
+        for<'de> T: Deserialize<'de>,
+    {
+        let response_values = self.request_batch(batch_params).await?;
+        let mut responses = Vec::with_capacity(response_values.len());
+
+        for (i, response_value) in response_values.into_iter().enumerate() {
+            let method = batch_params[i].0;
+            let params = &batch_params[i].1;
+
+            let response = serde_json::from_value(response_value).with_context(|| {
+                format!(
+                    "Unable to deserialize `{}` from JSON-RPC response for params: {:?}",
+                    method, params
+                )
+            })?;
+
+            responses.push(response);
+        }
+
+        Ok(responses)
+    }
+
+    /// Sends a JSON-RPC request and returns `request_id` and `response_channel`
+    async fn send_request(
+        &self,
+        method: &str,
+        params: &[Value],
+    ) -> Result<(String, Receiver<JsonRpcResponse>)> {
+        let id = self.unique_id.fetch_add(1, Ordering::Relaxed).to_string();
```

Comment:
I guess relaxed is fine here? Or would random 128-bit UUID make it more robust?

Reply:
I think this is fine. Didn't cause any problem until now.